Fix sidecar termination on migration: Introduces hooks v1alpha3 with Shutdown

victortoso opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
After migration, the sidecar is not being properly terminated. It does not receive a SIGTERM and it keeps running leaving a dangling Pod with `NotReady` state. The issue is well documented in https://github.com/kubevirt/kubevirt/issues/9838

This PR implements a Shutdown() mechanism to allow virt-launcher to inform the sidecar to exit.

**Which issue(s) this PR fixes**
Fixes #8395

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add v1alpha3 for hooks
Fix migration when using sidecars
```

victortoso commented:
/retest-required

prnaraya commented on cmd/example-hook-sidecar/smbios.go:
```diff
@@ -63,12 +67,41 @@ func (s infoServer) Info(ctx context.Context, params *hooksInfo.InfoParams) (*ho
+
+func (s v1alpha3Server) PreCloudInitIso(_ context.Context, _ *hooksV1alpha3.PreCloudInitIsoParams) (*hooksV1alpha3.PreCloudInitIsoResult, error) {
+	// Needs to be created to satisfy interface
+	panic("Not implemented. Not exported. Should not be called.")
```
I am not completely familiar with this, so you might have a good reason for doing this, but could you explain your reasoning behind panicking here and not implementing this? Thanks :)

victortoso replied:
This is not relevant anymore as `sidecar-shim` implements the `PreCloudInitIso`.

> could you explain your reasoning behind panicking here and not implementing this?

Yes. The smbios sidecar is only an example that changes the `baseBoardManufacturer` to `Radical Edward`. In order to do this, it needs to change libvirt's domain xml prior the VM start, using the `OnDefineDomain` hook callback.

It does not need other hooks.

We need to implement them for the sake of hook's interface. When I say we don't expose `PreCloudInitIso`, this is related to `Info` method, that provides to `virt-handler` or `hookManager` the functions/callbacks this sidecar expose.

prnaraya commented on pkg/hooks/manager.go:
```diff
@@ -260,75 +279,152 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
-				if err != nil {
-					return cloudInitData, fmt.Errorf("failed to marshal VMI spec: %v, err: %v", vmi, err)
-				}
+	dataToJSON := func() ([3][]byte, error) {
```
For readability sake, is it possible to move this func definition to its own separate helper function?

victortoso replied:
Fixed

prnaraya commented on pkg/hooks/manager.go:
```diff
@@ -260,75 +279,152 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
-				return resultData, nil
-			} else {
-				panic("Should never happen, version compatibility check is done during Info call")
+	validateResult := func(initData, noCloudSource []byte) (*cloudinit.CloudInitData, error) {
```
Same as above.

victortoso replied:
Fixed

prnaraya commented on pkg/hooks/manager.go:
```diff
@@ -260,75 +279,152 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
+		return resultData, nil
+	}
+
+	callbacks, found := m.CallbacksPerHookPoint[hooksInfo.PreCloudInitIsoHookPointName]
+	if !found {
+		return cloudInitData, nil
+	}
```
I am not sure it was necessary to change this from what it was originally here.

```suggestion
	if callbacks, found := m.CallbacksPerHookPoint[hooksInfo.PreCloudInitIsoHookPointName]; found {
```

victortoso replied:
Yes, this is a matter of preference, to return early when possible and reduce identation level. Considering I'm refactoring to add switch plus v1aplha3, I thought it would be fine to do, but not a necessity indeed.

prnaraya commented on pkg/hooks/manager.go:
```diff
@@ -260,75 +279,152 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
+			}
+			return validateResult(result.GetCloudInitData(), result.GetCloudInitNoCloudSource())
+		default:
+			log.Log.Errorf("Unsupported callback version: %s", callback.Version)
```
In conjunction with the above comment, is there a reason for changing this as well, rather than the original code that panicked? The statement here should still hold true.
```
panic("Should never happen, version compatibility check is done during Info call")
```

victortoso replied:
Yes, we should avoid panic. I see this as just another opportunity to avoid `virt-launcher` to panic due bad sidecar.

prnaraya commented on pkg/hooks/manager.go:
```diff
@@ -260,75 +279,152 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
 }
+
+func (m *hookManager) Shutdown() error {
+	callbacks, found := m.CallbacksPerHookPoint[hooksInfo.ShutdownHookPointName]
+	if !found {
+		return nil
+	}
```
Same as before in `PreCloudInitIso`, this can be consolidated to 
```
if callbacks, found := m.CallbacksPerHookPoint[hooksInfo.PreCloudInitIsoHookPointName]; found {
```

prnaraya commented on pkg/hooks/manager.go:
```diff
@@ -260,75 +279,152 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
+				return err
+			}
+		default:
+			log.Log.Errorf("Unsupported callback version: %s", callback.Version)
```
Same as in `PreCloudInitIso`

prnaraya reviewed (commented):
Thanks for addressing this issue! I've done an initial look through the PR and left a few inline comments and questions.

victortoso commented:
@prnaraya thanks for the review! jfyi, I did not forget about this PR but I thought I should first finish https://github.com/kubevirt/kubevirt/pull/9231 and https://github.com/kubevirt/kubevirt/pull/10320. They are not a hard requirement for this PR, just related and slightly time consuming ;)

I hope to get back to this next week!

victortoso commented on cmd/example-hook-sidecar/smbios.go:
```diff
@@ -38,12 +40,14 @@ import (
 const (
 	baseBoardManufacturerAnnotation = "smbios.vm.kubevirt.io/baseBoardManufacturer"
 	onDefineDomainLoggingMessage    = "Hook's OnDefineDomain callback method has been called"
+	onShutdownMessage               = "Hook's Shutdown's callback method has been called"
```
Fixed, thanks. Note that I'm moving the implementation from `smbios` to the `sidecar-shim` instead.

victortoso commented:
/cc @dharmit, in case you want to review.

If I did not misunderstood, this proposal of `Shutdown()` method would eventually not be needed anymore, with a proper fix in kubernetes.  See this [native sidecar containers](https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/), more specifically this quote:

> Pod termination continues to only depend on the main containers. An init container with a restartPolicy of Always (named a sidecar) won't prevent the pod from terminating after the main containers exit

As this is alpha, it might take some time so I think this PR is still useful.

toelke commented:
FWIW, we are running this PR in production.

I also think that once k8s 1.28 is required by kubevirt this feature could be dropped, but until then, it's massively helpful.

dharmit commented on pkg/hooks/info/info.go:
```diff
@@ -21,3 +21,4 @@ package info
 
 const OnDefineDomainHookPointName = "OnDefineDomain"
 const PreCloudInitIsoHookPointName = "PreCloudInitIso"
+const ShutdownHookPointName = "Shutdown"
```
nit: grouping these constants into a `const()` would improve readability, IMO.

victortoso replied:
Sure, I'll group them together.

dharmit reviewed (commented):
@victortoso do you plan to add functional tests in this PR or a separate one?

vasiliy-ul commented on pkg/hooks/manager.go:
```diff
@@ -252,83 +260,172 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
+			if err != nil {
+				log.Log.Reason(err).Error("Failed to run Shutdown")
+				return err
+			}
```
`defer conn.Close()` is missing here

victortoso replied:
Yes, thanks for catching that!

vasiliy-ul commented on cmd/sidecars/sidecar_shim.go:
```diff
@@ -241,6 +287,35 @@ func main() {
+	}
+
+	if err == nil {
+		server.Stop()
```
This should probably be `server.GracefulStop()`. Since the server now serves in a goroutine, there might be a race with a pending return of a result from the callbacks.

victortoso replied:
Yes, you are right.

victortoso commented on pkg/hooks/manager.go:
```diff
@@ -159,7 +161,13 @@ func processSideCarSocket(socketPath string) (*callBackClient, bool, error) {
 
-	if _, found := versionsSet[hooksV1alpha2.Version]; found {
+	if _, found := versionsSet[hooksV1alpha3.Version]; found {
+		return &callBackClient{
+			SocketPath:           socketPath,
+			Version:              hooksV1alpha3.Version,
+			subscribedHookPoints: info.GetHookPoints(),
+		}, false, nil
```
We could probably go with something like this (not yet tested)
```diff
diff --git a/pkg/hooks/manager.go b/pkg/hooks/manager.go
index 7610c1ed3..47b0ef3e4 100644
--- a/pkg/hooks/manager.go
+++ b/pkg/hooks/manager.go
@@ -161,29 +161,26 @@ func processSideCarSocket(socketPath string) (*callBackClient, bool, error) {
                versionsSet[version] = true
        }

-       if _, found := versionsSet[hooksV1alpha3.Version]; found {
-               return &callBackClient{
-                       SocketPath:           socketPath,
-                       Version:              hooksV1alpha3.Version,
-                       subscribedHookPoints: info.GetHookPoints(),
-               }, false, nil
-       } else if _, found := versionsSet[hooksV1alpha2.Version]; found {
-               return &callBackClient{
-                       SocketPath:           socketPath,
-                       Version:              hooksV1alpha2.Version,
-                       subscribedHookPoints: info.GetHookPoints(),
-               }, false, nil
-       } else if _, found := versionsSet[hooksV1alpha1.Version]; found {
-               return &callBackClient{
-                       SocketPath:           socketPath,
-                       Version:              hooksV1alpha1.Version,
-                       subscribedHookPoints: info.GetHookPoints(),
-               }, false, nil
-       } else {
-               return nil, false,
-                       fmt.Errorf("Hook sidecar does not expose a supported version. Exposed versions: %v, supported versions: %v",
-                               info.GetVersions(), []string{hooksV1alpha1.Version, hooksV1alpha2.Version})
+       // The order matters. We should match newer versions first.
+       supportedVersions := []string{
+               hooksV1alpha3.Version
+               hooksV1alpha2.Version,
+               hooksV1alpha1.Version,
        }
+
+       for _, version := range supportedVersions {
+               if _, found := versionsSet[version]; found {
+                       return &callBackClient{
+                               SocketPath:           socketPath,
+                               Version:              version,
+                               subscribedHookPoints: info.GetHookPoints(),
+                       }, false, nil
+               }
+       }
+
+       return nil, false,
+               fmt.Errorf("Hook sidecar does not expose a supported version. Exposed versions: %v, supported versions: %v",
+                       info.GetVersions(), []string{hooksV1alpha1.Version, hooksV1alpha2.Version})
 }
```

victortoso commented on pkg/hooks/manager.go:
```diff
@@ -347,9 +366,66 @@ func (m *hookManager) PreCloudInitIso(vmi *v1.VirtualMachineInstance, cloudInitD
 				return cloudInitData, err
 			}
 			return preCloudInitIsoValidateResult(cloudInitData.DataSource, result.GetCloudInitData(), result.GetCloudInitNoCloudSource())
+		case hooksV1alpha3.Version:
```
There is less we can do here. What I can see is factor out the `preCloudInitIsoDataToJSON()`.

The `conn` relies on the `callback` and the `client` and `result ` depend on hook version, so having their on block of logic is helpful.

Not tested:
```diff
diff --git a/pkg/hooks/manager.go b/pkg/hooks/manager.go
index 47b0ef3e4..880a301d8 100644
--- a/pkg/hooks/manager.go
+++ b/pkg/hooks/manager.go
@@ -333,15 +333,16 @@ func (m *hookManager) PreCloudInitIso(vmi *v1.VirtualMachineInstance, cloudInitD
        if !found {
                return cloudInitData, nil
        }
+
+       json, err := preCloudInitIsoDataToJSON(vmi, cloudInitData)
+       if err != nil {
+               log.Log.Reason(err).Error("Failed to run PreCloudInitIso")
+               return cloudInitData, err
+       }
+
        for _, callback := range callbacks {
                switch callback.Version {
                case hooksV1alpha2.Version:
-                       json, err := preCloudInitIsoDataToJSON(vmi, cloudInitData)
-                       if err != nil {
-                               log.Log.Reason(err).Error("Failed to run PreCloudInitIso")
-                               return cloudInitData, err
-                       }
-
                        conn, err := grpcutil.DialSocketWithTimeout(callback.SocketPath, 1)
                        if err != nil {
                                log.Log.Reason(err).Errorf(dialSockErr, callback.SocketPath)
@@ -364,12 +365,6 @@ func (m *hookManager) PreCloudInitIso(vmi *v1.VirtualMachineInstance, cloudInitD
                        }
                        return preCloudInitIsoValidateResult(cloudInitData.DataSource, result.GetCloudInitData(), result.GetCloudInitNoCloudSource())
                case hooksV1alpha3.Version:
-                       json, err := preCloudInitIsoDataToJSON(vmi, cloudInitData)
-                       if err != nil {
-                               log.Log.Reason(err).Error("Failed to run PreCloudInitIso")
-                               return cloudInitData, err
-                       }
-
                        conn, err := grpcutil.DialSocketWithTimeout(callback.SocketPath, 1)
                        if err != nil {
                                log.Log.Reason(err).Errorf(dialSockErr, callback.SocketPath)
```

victortoso commented:
> @victortoso do you plan to add functional tests in this PR or a separate one?

What I really wanted to do is a migration test with sidecar that fails without this PR. That in itself would be helpful in the future, when k8s fixes sidecar lifecycle. *But* I wasn't successful when I tried and time is a bit limited. I'll see if I make a simpler version to show that `Shutdown()` is indeed called.

Thank you for the reviews!

victortoso commented:
Fixed last suggestions, thanks @vasiliy-ul and @dharmit for the last round of reviews.

I've added a test to verify `sidecar-shim` is handling signals in the `refactor termination` commit, currently https://github.com/kubevirt/kubevirt/commit/e1a2debf3b580c1081fee9c40e549d335ebba689

I've added two tests, one with hooks v1alpha2 to show the actual bug I wanted to fix, failing to terminate the sidecar container on source pod over migration; and another test, using v1alpha3 to show that this PR works as expected. Both in the last commit `validate behavior on migration` or https://github.com/kubevirt/kubevirt/commit/628602d43a1b4f7ba0c04aef168e29e13b1e5583

/cc @alicefr

victortoso commented:
/retest-required

victortoso commented:
/retest-required

dharmit reviewed (commented):
More of open questions/comments than request for changes.

dharmit commented:
@victortoso of everything I have put in above review, only https://github.com/kubevirt/kubevirt/pull/10046#discussion_r1397064426 might need to be changed. Everything else is open question, and a couple are nitpicks.

victortoso commented on pkg/hooks/manager.go:
```diff
@@ -348,7 +348,7 @@ func (m *hookManager) PreCloudInitIso(vmi *v1.VirtualMachineInstance, cloudInitD
 			return preCloudInitIsoValidateResult(cloudInitData.DataSource, result.GetCloudInitData(), result.GetCloudInitNoCloudSource())
 		default:
-			panic("Should never happen, version compatibility check is done during Info call")
+			log.Log.Errorf("Unsupported callback version: %s", callback.Version)
```
The commit can be dropped if we prefer to panic. If we squash it, it would mean we changed a particular behavior that might not be desirable together with other commit that does something else. It would be different if I have a few commits doing "do not panic".

That said, I'd be more inclined to drop this commit instead of squashing it as it is tangent to the goal of the PR.

dharmit replied:
> I'd be more inclined to drop this commit instead of squashing it as it is tangent to the goal of the PR.

Up to you if you want to drop or keep, but I agree with what you said in https://github.com/kubevirt/kubevirt/pull/10046#discussion_r1374833358 - A bad sidecar shouldn't panic virt-launcher

victortoso commented on tests/vmi_hook_sidecar_test.go:
```diff
@@ -218,6 +219,40 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
 			})
 		})
 
+		Context("with sidecar-shim", func() {
```
I'm considering doing a small refactor on the sidecar tests as a follow up. We can probably reduce the amount of tests without losing code coverage. In general, our goal is to test hooks behavior and we now should be able to do that 100% with sidecar-shim.

So, yes, let's do that but as a follow up if you don'´t mind.

dharmit replied:
Sure thing. Fine by me. 👍🏽

victortoso commented on tests/vmi_hook_sidecar_test.go:
```diff
@@ -218,6 +219,40 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
+				Expect(err).ToNot(HaveOccurred())
+				libwait.WaitForSuccessfulVMIStart(vmi)
+
+				time.Sleep(100 * time.Millisecond)
```
sigh, this is a leftover. Thanks for catching.
Just to explain it, I had it before I start using `Eventually` which is basically the same thing.

I'll remove it shortly.

victortoso commented on tests/vmi_hook_sidecar_test.go:
```diff
@@ -218,6 +219,40 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
 
+		Context("with sidecar-shim", func() {
+			It("should receive Terminal signal on VMI deletion", func() {
+				vmi, err = virtClient.VirtualMachineInstance(testsuite.GetTestNamespace(nil)).Create(context.Background(), vmi)
+				Expect(err).ToNot(HaveOccurred())
+				libwait.WaitForSuccessfulVMIStart(vmi)
```
Thanks, I'll fix this too.

victortoso commented on tests/vmi_hook_sidecar_test.go:
```diff
@@ -218,6 +219,40 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
+				libwait.WaitForSuccessfulVMIStart(vmi)
+
+				time.Sleep(100 * time.Millisecond)
+				err = virtClient.VirtualMachineInstance(vmi.Namespace).Delete(context.Background(), vmi.Name, &metav1.DeleteOptions{})
```
Ack.

victortoso commented on tests/vmi_hook_sidecar_test.go:
```diff
@@ -218,6 +219,40 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
+						}).
+						DoRaw(context.Background())
+					Expect(err).ToNot(HaveOccurred())
+					return strings.Contains(string(logsRaw), "sidecar-shim received signal: terminated")
```
I agree, actually I reopened https://github.com/kubevirt/kubevirt/issues/3764#issuecomment-1814026363 yesterday for that reason.

We need to write the code to make it possible though.

victortoso replied:
> In this case, however, I wonder what that condition could be. Maybe we could validate if the initial vmiPod gets successfully deleted/terminated?

The main goal for me is to validate that migration works and be able to show (as in an actual reproducer) that it doesn't.

There are several types of testing, I'm not super knowledgeable in all of them. Testing that `Shutdown()` is called would be a simpler version of what I did, but what I did I think is more important at the moment.

All in all, we can rethink this in a follow up.

victortoso commented on cmd/sidecars/sidecar_shim.go:
```diff
@@ -90,6 +101,39 @@ func (s infoServer) Info(ctx context.Context, params *hooksInfo.InfoParams) (*ho
+	done chan struct{}
+}
+
+func (s v1Alpha3Server) OnDefineDomain(_ context.Context, params *hooksV1alpha3.OnDefineDomainParams) (*hooksV1alpha3.OnDefineDomainResult, error) {
+	log.Log.Info(onDefineDomainLoggingMessage)
+	newDomainXML, err := runOnDefineDomain(params.GetVmi(), params.GetDomainXML())
+	if err != nil {
+		log.Log.Reason(err).Error("Failed OnDefineDomain")
+		return nil, err
+	}
+	return &hooksV1alpha3.OnDefineDomainResult{
+		DomainXML: newDomainXML,
+	}, nil
+}
```
This was asked before, this is my answer:
https://github.com/kubevirt/kubevirt/pull/9231#discussion_r1255407468

Instead, the path forward is deprecating old v1alpha1 and removing it from the codebase and keep moving the tests to a more recent version. This is something I plan to propose later on as well.

dharmit replied:
> Instead, the path forward is deprecating old v1alpha1 and removing it from the codebase and keep moving the tests to a more recent version. This is something I plan to propose later on as well.

Makes sense. And since it's alpha version API, we can even deprecate all but last? Anyway, agree on doing this later. 👍🏽

victortoso commented on tests/vmi_hook_sidecar_test.go:
```diff
@@ -251,6 +255,54 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
+				Expect(err).ToNot(HaveOccurred())
+				Expect(exists).To(BeTrue())
+
+				sourceName := sourcePod.GetObjectMeta().GetName()
```
ack

victortoso commented on tests/vmi_hook_sidecar_test.go:
```diff
@@ -251,6 +255,54 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
+				Expect(err).ToNot(HaveOccurred())
+				Expect(exists).To(BeTrue())
+
+				Expect(sourcePod.Name).ToNot(Equal(targetPod.Name))
```
Over a namespace I think name should be unique. But UUID is more sure-proof so let's change this as well.

Thanks for the review!

victortoso commented:
Shameless ping! With this last version, I've added tests to validate (shim) sidecars are receiving/handling signals normally and also a test to show the bug #8395 and #9838 prior to v1alpha3 and that the bug is solved by this PR.

@alicefr @vasiliy-ul @0xFelix

vasiliy-ul commented on pkg/hooks/manager.go:
```diff
@@ -159,23 +161,26 @@ func processSideCarSocket(socketPath string) (*callBackClient, bool, error) {
+
+	return nil, false,
+		fmt.Errorf("Hook sidecar does not expose a supported version. Exposed versions: %v, supported versions: %v",
+			info.GetVersions(), []string{hooksV1alpha1.Version, hooksV1alpha2.Version})
```
The error message does not include `v3alpha` as `supported versions`
```suggestion
			info.GetVersions(), supportedVersions)
```

victortoso replied:
Thanks! Fixed it.

vasiliy-ul commented on pkg/hooks/manager.go:
```diff
@@ -252,83 +257,169 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
+	empty := [3][]byte{}
+	vmiJSON, err := json.Marshal(vmi)
+	if err != nil {
+		return empty, fmt.Errorf("failed to marshal VMI spec: %v, err: %v", vmi, err)
```
nit: vmi is a big struct. Printing it may result in flooded log. I would just return the error message. The same might be considered when printing `cloudInitData` in the code below. But feel free to ignore if you disagree.

victortoso replied:
That's how it was before, I just moved the code to make introducing the new v1alpha3 easier but I agree with you that that error is enough here. I'll fix and add a note in the commit log about the change.

victortoso replied:
Not tested yet, just to clarify the change.

```diff
diff --git a/pkg/hooks/manager.go b/pkg/hooks/manager.go
index 36ced5aae..8ee154607 100644
--- a/pkg/hooks/manager.go
+++ b/pkg/hooks/manager.go
@@ -279,7 +279,7 @@ func preCloudInitIsoDataToJSON(vmi *v1.VirtualMachineInstance, cloudInitData *cl
        empty := [3][]byte{}
        vmiJSON, err := json.Marshal(vmi)
        if err != nil {
-               return empty, fmt.Errorf("failed to marshal VMI spec: %v, err: %v", vmi, err)
+               return empty, fmt.Errorf("failed to marshal VMI spec, err: %v", err)
        }

        // To be backward compatible to sidecar hooks still expecting to receive the cloudinit data as a
@@ -297,7 +297,7 @@ func preCloudInitIsoDataToJSON(vmi *v1.VirtualMachineInstance, cloudInitData *cl

        cloudInitDataJSON, err := json.Marshal(cloudInitData)
        if err != nil {
-               return empty, fmt.Errorf("failed to marshal CloudInitData: %v, err: %v", cloudInitData, err)
+               return empty, fmt.Errorf("failed to marshal CloudInitData, err: %v", err)
        }
```

vasiliy-ul commented on tests/vmi_hook_sidecar_test.go:
```diff
@@ -218,6 +223,87 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
+				sourcePodName := sourcePod.GetObjectMeta().GetName()
+				sourcePodUID := sourcePod.GetObjectMeta().GetUID()
+
+				migration := tests.NewRandomMigration(vmi.Name, vmi.Namespace)
```
nit:
```suggestion
				migration := tests.NewRandomMigration(vmi.Name, testsuite.GetTestNamespace(vmi))
```

victortoso replied:
Fixed too.

vasiliy-ul reviewed (commented):
Looks good from my perspective. No issues spotted, just  couple of nits.

/approve

> a test to show the bug https://github.com/kubevirt/kubevirt/issues/8395 and https://github.com/kubevirt/kubevirt/issues/9838 prior to v1alpha3 and that the bug is solved by this PR.

:+1:

victortoso commented:
/retest

0xFelix commented on pkg/hooks/manager.go:
```diff
@@ -252,83 +257,169 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
 	return domainSpecXML, nil
 }
 
+func preCloudInitIsoDataToJSON(vmi *v1.VirtualMachineInstance, cloudInitData *cloudinit.CloudInitData) ([3][]byte, error) {
```
Can't you return individual values (or ptrs to them) instead of the byte slice?

victortoso replied:
Yes, we can do that... I'll change it.

0xFelix commented on pkg/hooks/manager.go:
```diff
@@ -252,83 +257,169 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
+func preCloudInitIsoValidateResult(dataSource cloudinit.DataSourceType, initData, noCloudSource []byte) (*cloudinit.CloudInitData, error) {
+	var resultData *cloudinit.CloudInitData
+	err := json.Unmarshal(initData, &resultData)
+	if err != nil {
```
nit: Inline check

victortoso replied:
I'll fix the nits in the next iteration.

0xFelix commented on pkg/hooks/manager.go:
```diff
@@ -252,83 +257,169 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
+		// Be backwards compatible for hook sidecars still working on CloudInitNoCloudSource objects instead of CloudInitData
+		var resultNoCloudSourceData *v1.CloudInitNoCloudSource
+		err = json.Unmarshal(noCloudSource, &resultNoCloudSourceData)
+		if err != nil {
```
nit: Inline check

0xFelix commented on pkg/hooks/manager.go:
```diff
@@ -252,83 +257,169 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
+			ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+			defer cancel()
+
+			_, err = client.Shutdown(ctx, &hooksV1alpha3.ShutdownParams{})
```
nit: Inline check?

0xFelix commented on tests/vmi_hook_sidecar_test.go:
```diff
@@ -218,6 +223,87 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
+				Eventually(func() bool {
+					vmiPod, exists, err := getVMIPod(vmi)
+					if err != nil {
+						Expect(err).NotTo(HaveOccurred(), "must be able to retrieve the VMI virt-launcher pod")
```
This is weird, remove guard around `Expect(err).NotTo(HaveOccurred())`?

0xFelix replied:
Also, if you Expect inside Eventually you most likely want to do it like this:

```
Eventually(func(g Gomega) {
    g.Expect(err).ToNot(HaveOccured())
...
```

Otherwise you fail the whole test instead of running the Eventually func multiple times, what is probably what you intended to do.

0xFelix commented on tests/vmi_hook_sidecar_test.go:
```diff
@@ -218,6 +223,87 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
+					Expect(err).ToNot(HaveOccurred())
+					return strings.Contains(string(logsRaw), "sidecar-shim received signal: terminated")
+				}, 30*time.Second, time.Second).Should(
+					BeTrue(),
```
If you expect inside the Eventually func (`g Gomega`), you can match on `Succeeded()` here instead.

victortoso replied:
Thanks, I'm still learning this test framework. I'll change it.

0xFelix commented on tests/vmi_hook_sidecar_test.go:
```diff
@@ -218,6 +223,87 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
+						metav1.ListOptions{
+							FieldSelector: fields.ParseSelectorOrDie("metadata.name=" + sourcePodName).String(),
+						})
+					Expect(err).ToNot(HaveOccurred())
```
Same as above.

0xFelix commented on tests/vmi_hook_sidecar_test.go:
```diff
@@ -218,6 +223,87 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
+						}).
+						DoRaw(context.Background())
+					Expect(err).ToNot(HaveOccurred())
+					return strings.Contains(string(logsRaw), "sidecar-shim received signal: terminated")
```
Do this with `g.Expect(string(logsRaw)).To(ContainSubstring("..."))` instead?

victortoso commented:
Hi @0xFelix please double check the test to see if that's what you expected. All the changes on tests were [here](https://github.com/kubevirt/kubevirt/pull/10046/commits/d97cba19b55392153c5bdf8879c87c689b46d09e#diff-fe9ff27d97b7be1c5a5c7a3d04f614eda6c6c3afbeb8a9d38bdf90955917c69a):

It worked locally. I think it turned out much better than it was indeed.

victortoso commented:
/retest

victortoso commented:
/retest

0xFelix commented on pkg/hooks/manager.go:
```diff
@@ -252,83 +257,165 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
+func preCloudInitIsoDataToJSON(vmi *v1.VirtualMachineInstance, cloudInitData *cloudinit.CloudInitData) ([]byte, []byte, []byte, error) {
+	vmiJSON, err := json.Marshal(vmi)
+	if err != nil {
+		return []byte{}, []byte{}, []byte{}, fmt.Errorf("failed to marshal VMI spec, err: %v", err)
```
You can return `nil` instead of `[]byte{}`.

victortoso replied:
```diff
--- a/pkg/hooks/manager.go
+++ b/pkg/hooks/manager.go
@@ -262,7 +262,7 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
 func preCloudInitIsoDataToJSON(vmi *v1.VirtualMachineInstance, cloudInitData *cloudinit.CloudInitData) ([]byte, []byte, []byte, error) {
        vmiJSON, err := json.Marshal(vmi)
        if err != nil {
-               return []byte{}, []byte{}, []byte{}, fmt.Errorf("failed to marshal VMI spec, err: %v", err)
+               return nil, nil, nil, fmt.Errorf("failed to marshal VMI spec, err: %v", err)
        }

        // To be backward compatible to sidecar hooks still expecting to receive the cloudinit data as a
@@ -275,12 +275,12 @@ func preCloudInitIsoDataToJSON(vmi *v1.VirtualMachineInstance, cloudInitData *cl
        }
        cloudInitNoCloudSourceJSON, err := json.Marshal(cloudInitNoCloudSource)
        if err != nil {
-               return []byte{}, []byte{}, []byte{}, fmt.Errorf("failed to marshal CloudInitNoCloudSource: %v, err: %v", cloudInitNoCloudSource, err)
+               return nil, nil, nil, fmt.Errorf("failed to marshal CloudInitNoCloudSource: %v, err: %v", cloudInitNoCloudSource, err)
        }

        cloudInitDataJSON, err := json.Marshal(cloudInitData)
        if err != nil {
-               return []byte{}, []byte{}, []byte{}, fmt.Errorf("failed to marshal CloudInitData, err: %v", err)
+               return nil, nil, nil, fmt.Errorf("failed to marshal CloudInitData, err: %v", err)
 ```
 
 To be squashed in the first commit

0xFelix commented on pkg/hooks/manager.go:
```diff
@@ -252,83 +257,165 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
+	}
+	cloudInitNoCloudSourceJSON, err := json.Marshal(cloudInitNoCloudSource)
+	if err != nil {
+		return []byte{}, []byte{}, []byte{}, fmt.Errorf("failed to marshal CloudInitNoCloudSource: %v, err: %v", cloudInitNoCloudSource, err)
```
Same.

0xFelix commented on pkg/hooks/manager.go:
```diff
@@ -252,83 +257,165 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
+
+	cloudInitDataJSON, err := json.Marshal(cloudInitData)
+	if err != nil {
+		return []byte{}, []byte{}, []byte{}, fmt.Errorf("failed to marshal CloudInitData, err: %v", err)
```
Same.

0xFelix commented on tests/vmi_hook_sidecar_test.go:
```diff
@@ -218,6 +222,84 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
+				targetPodUID := targetPod.GetObjectMeta().GetUID()
+				Expect(sourcePodUID).ToNot(Equal(targetPodUID))
+
+				Eventually(func() bool {
```
Introduce `g Gomega` here too?

victortoso replied:
```diff
-                               Eventually(func() bool {
+                               Eventually(func(g Gomega) {
                                        pods, err := virtClient.CoreV1().Pods("").List(
                                                context.Background(),
                                                metav1.ListOptions{
                                                        FieldSelector: fields.ParseSelectorOrDie("metadata.name=" + sourcePodName).String(),
                                                })
-                                       Expect(err).ToNot(HaveOccurred())
-                                       Expect(pods.Items).To(HaveLen(1))
+                                       g.Expect(err).ToNot(HaveOccurred())
+                                       g.Expect(pods.Items).To(HaveLen(1))
                                        computeTerminated := false
                                        sidecarTerminated := false
                                        for _, container := range pods.Items[0].Status.ContainerStatuses {
@@ -291,8 +291,9 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
                                                        sidecarTerminated = hasTerminated
                                                }
                                        }
-                                       return computeTerminated && (sidecarTerminated == sidecarShouldTerminate)
-                               }, 30*time.Second, 1*time.Second).Should(BeTrue())
+                                       g.Expect(computeTerminated).To(BeTrue())
+                                       g.Expect(sidecarTerminated).To(Equal(sidecarShouldTerminate))
+                               }, 30*time.Second, 1*time.Second).Should(Succeed())
 ```
 
 To be squashed in the last commit

victortoso commented:
/retest

0xFelix reviewed (approved):
Thanks! :)

/lgtm

brianmcarey commented:
/test all
